goguen/formal-spec/utxo.tex

```diff
@@ -746,17 +697,26 @@ \subsection{Witnessing}
   \begin{align*}
     & \hspace{-1cm}\fun{scriptsNeeded} \in \UTxO \to \GoguenTx \to \powerset (\ScriptPurpose \times \ScriptHash) \\
     & \hspace{-1cm}\fun{scriptsNeeded}~\var{utxo}~\var{tx} = \\
-    & ~~\{ (\var{i}, \fun{validatorHash}~a) \mid i \mapsto (a, \wcard) \in \var{utxo},\\
-    & ~~~~~i\in\fun{txinsScript}~{(\fun{txins~\var{txb}})}~{utxo}\} \\
+    & ~~\{ (\var{i}, \fun{validatorHash}~a) \mid i \mapsto (a, \wcard, \wcard) \in \var{utxo},
+      i\in\fun{txinsScript}~{(\fun{txins~\var{txb}})}~{utxo}\} \\
     \cup & ~~\{ (\var{a}, \fun{stakeCred_{r}}~\var{a}) \mid a \in \dom (\AddrRWDScr
            \restrictdom \fun{txwdrls}~\var{txb}) \} \\
     \cup & ~~\{ (\var{cert}, \var{c}) \mid \var{cert} \in (\DCertDeleg \cup \DCertDeRegKey)\cap\fun{txcerts}~(\txbody{tx}), \\
     & ~~~~~~\var{c} \in \cwitness{cert} \cap \AddrScr\} \\
       \cup & ~~\{ (\var{pid}, \var{pid}) \mid \var{pid} \in \supp~(\fun{forge}~\var{txb}) \} \\
     & \where \\
     & ~~~~~~~ \var{txb}~=~\txbody{tx}
+    \nextdef
+    & \hspace{-1cm}\fun{hashSD} \in \TxWitness \to \ScriptDataHash^? \\
+    & \hspace{-1cm}\fun{hashSD}~\var{txw}~=~
+      \begin{cases}
+        \Nothing & (\var{sc}, \var{dat}, \var{rdmr}) ~=~ (\emptyset, \emptyset, \emptyset) \\
+        \fun{hash}~(\var{sc}, \var{dat}, \var{rdmr}) & \text{otherwise}
```

Comment:
TxWitness includes the signatures too - i think you want `(sigs, sc, dat, rdmr)` and ignore sigs in the hashing

Reply:
Note that below there's a where clause that matches the witnesses against that, ignoring the signatures